### ISSUE TITLE
Potential concurrency issue for AdjustInventoryOrderChangedEventHandler

### ISSUE THREAD

**Issue by leelu611** [VirtoCommerce/vc-module-order#69]
Hi there,

We find that there might be potential concurrency issue for the inventory adjustment if there are multiple orders are placed with the same SKU at the same time. This seems to be caused by the following code blocks in AdjustInventoryOrderChangedEventHandler.
Basically the handler tries to get the original inventory infos first and then do the adjustments and commit. However, if there multiple events triggerd for the same SKU at the same time, some of them might get the original in stock quantity before others doing the updates, rather than the updated value. This happens quite often if we have very high load or different order creation entries.

```csharp
            var inventoryAdjustments = new HashSet<InventoryInfo>();
            //Load all inventories records for all changes and old order items
            var inventoryInfos = _inventoryService.GetProductsInventoryInfos(origLineItems.Select(x => x.ProductId).Concat(changedLineItems.Select(x => x.ProductId)).Distinct().ToArray());
            changedLineItems.CompareTo(origLineItems, EqualityComparer<LineItem>.Default, (state, changed, orig) => { AdjustInventory(inventoryInfos, inventoryAdjustments, customerOrder, state, changed, orig); });
            //Save inventories adjustments
            if (inventoryAdjustments != null)
            {
                _inventoryService.UpsertInventories(inventoryAdjustments);
            }
```

We think a quick fix might be to move the inventory adjustment into a sequential job queue, e.g. using Hangfire job and DisableConcurentExecution to ensure the inventory are adjusted sequentially.

Please help to verify the issue.

**Comment by tatarincev** [comment]
@leelu611 I completely confirm what the current implementation has this issue. So there are a some solutions, each of them have a pros and cons
1. Use the **Hangfire** job for this purposes as you suggest (don't guaranteed the data consistency the background job might be canceled)
2. Use critical section (lock statement) in the **AdjustInventoryOrderChangedEventHandler**  for who.e  handler (doesn't work for multiple platform instances)
3. Define the special concurrent safe method for inventory adjustment on the inventory module side instead of entire inventory update operation (via registration the new adjustment operation that can be used for actual inventory remains calculation)
4. Calculate an actual inventory from placed orders with the special states (required to major changes with inventory logic) 

The most easiest for me is the number 1. do an inventory adjustment in the Hangfire background job which will scheduled directly in the handler `Handle` method.
More correct implementation is probably number 3 to define special command for inventory adjustment on the inventory module side.

**Comment by tatarincev** [comment]
@avas the future changes should looks like this code, but these code doesn't work because Hangfire uses serialization for action arguments, and missing parameter types cannot be deserialized due to the missed default constructors. Seems need to refactor this handler to allow use Hagfire jobs for inventory adjustments without an issue related to  concurrent/parallels updates. 
![image](https://user-images.githubusercontent.com/7566324/47721811-997b6100-dc59-11e8-9e42-a68a141d9959.png)

**Comment by leelu611** [comment]
@tatarincev Thank you very much! Indeed in our case, the best option currently is the first one as we do have setup a multiple instances platform. Due to the deserialization of the HF job, probably we need to first calculate the product ids and delta quantity for each order and pass this simple object to the HF job.

**Comment by avas** [comment]
@tatarincev Well, the problem lies a bit deeper. To pass old and new orders to the job, we'll need to serialize and deserialize some `CustomerOrder` instances. It contains some implementations of `IOperation`, so it requires a `PolymorphicOperationJsonConverter` to properly deserialize these implementations. However, Hangfire seems to use its own serialization settings, and it doesn't currently use JSON converters registered by the platform and modules - I think this isn't right.

So, to solve it, I modified Hangfire settings in the platform startup code:
```csharp
public class HangfireLauncher
{
    // ...

    public void ConfigureOwin(IAppBuilder app, IUnityContainer container)
    {
        // ...

        // Configure Hangfire to use platform serialization settings for job parameters serialization.
        // This will force Hangfire to use all PolymorphicJsonConverters configured by platform and modules.
        var httpConfiguration = container.Resolve<HttpConfiguration>();
        var serializerSettings = httpConfiguration.Formatters.JsonFormatter.SerializerSettings;
        JobHelper.SetSerializerSettings(serializerSettings);
    }

    // ...
}
```
After that, passing `CustomerOrders` to the job seems to work. But this change can potentially break deserialization of already queued jobs with complex parameters. So, is it safe to change serialization settings for Hangfire? Can I use this solution?

**Comment by tatarincev** [comment]
I think we need refractor current method logic to adjust inventory in method that can be easy run within background job with simple formal parameters

**Comment by avas** [comment]
@tatarincev OK, got it! I've tested this attempt, and it appears to work without modifying Hangfire configuration.

So, I did the following:
1. Replaced `GenericChangeEntry<CustomerOrder>` with actual data used during inventory adjustment (`storeId` for order, order shipments, old and new line items...).
2. Encapsulated this data to a custom type and passed it to the background job.

I will create a PR with these changes soon.
